bug:  Mock Provider - Union

Please read our [contributor guide](https://github.com/Netflix/dgs-framework/blob/master/CONTRIBUTING.md) before
creating an issue.

## Expected behavior
Application should start without exception

## Actual behavior
Exception in thread "main" org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'dgsQueryExecutor' defined in class path resource [com/netflix/graphql/dgs/autoconfig/DgsAutoConfiguration.class]: Unsatisfied dependency expressed through method 'dgsQueryExecutor' parameter 1: Error creating bean with name 'schema' defined in class path resource [com/netflix/graphql/dgs/autoconfig/DgsAutoConfiguration.class]: Failed to instantiate [graphql.schema.GraphQLSchema]: Factory method 'schema' threw exception with message: class graphql.schema.GraphQLInterfaceType cannot be cast to class graphql.schema.GraphQLObjectType (graphql.schema.GraphQLInterfaceType and graphql.schema.GraphQLObjectType are in unnamed module of loader 'app')
                at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800)
                at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:550)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520)
                at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326)
                at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
                at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324)
                at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
                at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:973)
                at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:942)
                at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:608)
                at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
                at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:734)
                at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:436)
                at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
                at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
                at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'schema' defined in class path resource [com/netflix/graphql/dgs/autoconfig/DgsAutoConfiguration.class]: Failed to instantiate [graphql.schema.GraphQLSchema]: Factory method 'schema' threw exception with message: class graphql.schema.GraphQLInterfaceType cannot be cast to class graphql.schema.GraphQLObjectType (graphql.schema.GraphQLInterfaceType and graphql.schema.GraphQLObjectType are in unnamed module of loader 'app')
                at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:659)
                at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:647)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560)
                at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520)
                at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326)
                at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
                at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324)
                at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
                at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
                at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1417)
                at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1337)
                at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:888)
                at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
                ... 19 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [graphql.schema.GraphQLSchema]: Factory method 'schema' threw exception with message: class graphql.schema.GraphQLInterfaceType cannot be cast to class graphql.schema.GraphQLObjectType (graphql.schema.GraphQLInterfaceType and graphql.schema.GraphQLObjectType are in unnamed module of loader 'app')
                at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171)
                at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:655)
                ... 33 more
Caused by: java.lang.ClassCastException: class graphql.schema.GraphQLInterfaceType cannot be cast to class graphql.schema.GraphQLObjectType (graphql.schema.GraphQLInterfaceType and graphql.schema.GraphQLObjectType are in unnamed module of loader 'app')
                at com.netflix.graphql.mocking.MockGraphQLVisitor.visitGraphQLFieldDefinition(MockGraphQLVisitor.kt:62)
                at graphql.schema.GraphQLFieldDefinition.accept(GraphQLFieldDefinition.java:201)
                at graphql.schema.SchemaTraverser$TraverserDelegateVisitor.enter(SchemaTraverser.java:109)
                at graphql.util.Traverser.traverse(Traverser.java:144)
                at graphql.schema.SchemaTraverser.doTraverse(SchemaTraverser.java:96)
                at graphql.schema.SchemaTraverser.depthFirst(SchemaTraverser.java:86)
                at graphql.schema.SchemaTraverser.depthFirst(SchemaTraverser.java:79)
                at graphql.schema.SchemaTraverser.depthFirst(SchemaTraverser.java:75)
                at com.netflix.graphql.mocking.DgsSchemaTransformer.transformSchema(DgsSchemaTransformer.kt:35)
                at com.netflix.graphql.mocking.DgsSchemaTransformer.transformSchemaWithMockProviders(DgsSchemaTransformer.kt:29)
                at com.netflix.graphql.dgs.internal.DgsSchemaProvider.computeSchema(DgsSchemaProvider.kt:210)
                at com.netflix.graphql.dgs.internal.DgsSchemaProvider.schema(DgsSchemaProvider.kt:132)
                at com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration.schema(DgsAutoConfiguration.kt:219)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:568)
                at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:139)

## Steps to reproduce

Schema: 
type Person {
    id: ID
    name: Name
    salary: Float
    married: Boolean
    address: Address
}

type Name {
    firstname: String
    middleName: String
    lastName: String
}

type Address {
    houseNumber: Int
    street: String
    state: String
    country: String
}

type Query {
    person(name: String): PersonResult
}
interface Error {
    code: String!
    reason: String
    message: String
}

type PersonError implements Error {
    code: String!
    reason: String
    message: String
}

union PersonResult = Person  | PersonError


Mock Java class:

@Component

public class PersonDataProvider implements MockProvider {

    @Override
    public Map<String, Object> provide() {
        var map = new HashMap<String, Object>();
        map.put("person",null);
        return map;
    }


}

Additional info:

Spring boot Parent - spring-boot-starter-parent - 3.1.0
graphql-dgs-platform-dependencies - 7.3.3
Java 17

_Note: A test case would be highly appreciated, but we understand that's not always possible_

--- a/graphql-dgs-mocking/build.gradle.kts
+++ b/graphql-dgs-mocking/build.gradle.kts
@@ -18,4 +18,6 @@ dependencies {
     api("com.graphql-java:graphql-java")
     implementation("net.datafaker:datafaker:2.+")
     implementation("org.slf4j:slf4j-api")
+
+    testImplementation("org.assertj:assertj-core")
 }

--- a/graphql-dgs-mocking/src/main/kotlin/com/netflix/graphql/mocking/DgsSchemaTransformer.kt
+++ b/graphql-dgs-mocking/src/main/kotlin/com/netflix/graphql/mocking/DgsSchemaTransformer.kt
@@ -17,31 +17,26 @@
 package com.netflix.graphql.mocking
 
 import graphql.schema.*
-import java.util.*
 
 class DgsSchemaTransformer {
 
     fun transformSchemaWithMockProviders(schema: GraphQLSchema, mockProviders: Set<MockProvider>): GraphQLSchema {
-        val mockConfig = HashMap<String, Any>()
-
+        val mockConfig = mutableMapOf<String, Any>()
         mockProviders.forEach { p -> mockConfig.putAll(p.provide()) }
-
         return transformSchema(schema, mockConfig)
     }
 
     fun transformSchema(schema: GraphQLSchema, mockConfig: Map<String, *>): GraphQLSchema {
-        val mockFetchers = HashMap<FieldCoordinates, DataFetcher<*>>()
+        val mockFetchers = mutableMapOf<FieldCoordinates, DataFetcher<*>>()
         val graphQLTypeVisitorStub = MockGraphQLVisitor(mockConfig, mockFetchers)
-        SchemaTraverser().depthFirst(graphQLTypeVisitorStub, schema.getType("Query"))
 
-        return schema.transform { b ->
-            val newCodeRegistry = GraphQLCodeRegistry.newCodeRegistry(schema.codeRegistry)
-            mockFetchers.forEach { (coordinates, dataFetcher) -> newCodeRegistry.dataFetcher(coordinates, dataFetcher) }
+        SchemaTraverser().depthFirstFullSchema(graphQLTypeVisitorStub, schema)
 
-            b.codeRegistry(
-                newCodeRegistry
-                    .build()
-            )
+        return schema.transform { schemaBuilder ->
+            val newCodeRegistry = schema.codeRegistry.transform { codeRegistryBuilder ->
+                mockFetchers.forEach { (coordinates, dataFetcher) -> codeRegistryBuilder.dataFetcher(coordinates, dataFetcher) }
+            }
+            schemaBuilder.codeRegistry(newCodeRegistry)
         }
     }
 }

--- a/graphql-dgs-mocking/src/main/kotlin/com/netflix/graphql/mocking/MockGraphQLVisitor.kt
+++ b/graphql-dgs-mocking/src/main/kotlin/com/netflix/graphql/mocking/MockGraphQLVisitor.kt
@@ -16,6 +16,8 @@
 
 package com.netflix.graphql.mocking
 
+import graphql.Scalars
+import graphql.introspection.Introspection
 import graphql.schema.*
 import graphql.util.TraversalControl
 import graphql.util.TraverserContext
@@ -24,64 +26,103 @@ import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 
 class MockGraphQLVisitor(private val mockConfig: Map<String, Any?>, private val mockFetchers: MutableMap<FieldCoordinates, DataFetcher<*>>) : GraphQLTypeVisitorStub() {
-    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)
-    private val providedRoots: MutableList<String?> = mutableListOf()
+    companion object {
+        private val logger: Logger = LoggerFactory.getLogger(MockGraphQLVisitor::class.java)
+    }
+    private val additionalObjectTypes = mutableSetOf<GraphQLObjectType>()
+    private val providedRoots: MutableList<String> = mutableListOf()
     private val faker = Faker()
 
-    override fun visitGraphQLFieldDefinition(node: GraphQLFieldDefinition?, context: TraverserContext<GraphQLSchemaElement>?): TraversalControl {
-        val pathForNode = getPathForNode(context?.parentNodes, node)
+    override fun visitGraphQLFieldDefinition(node: GraphQLFieldDefinition, context: TraverserContext<GraphQLSchemaElement>): TraversalControl {
+        val parentNode = context.parentNode as GraphQLFieldsContainer
+        if (Introspection.isIntrospectionTypes(parentNode)) {
+            return TraversalControl.CONTINUE
+        }
 
-        if (mockConfig.keys.any { pathForNode?.startsWith(it) == true } && !providedRoots.any { pathForNode?.startsWith(it!!) == true && pathForNode.count { c -> c == '.' } != it?.count { c -> c == '.' } }) {
-            val type = if (node?.type is GraphQLNonNull) (node.type as GraphQLNonNull).wrappedType else node?.type
+        val pathForNode = getPathForNode(context.parentNodes, node)
+
+        if (parentNode in additionalObjectTypes || (mockConfig.keys.any { pathForNode.startsWith(it) } && !providedRoots.any { pathForNode.startsWith(it) && pathForNode.count { c -> c == '.' } != it.count { c -> c == '.' } })) {
+            if (parentNode is GraphQLInterfaceType) {
+                val schema = context.getVarFromParents(GraphQLSchema::class.java)
+                for (objType in schema.getImplementations(parentNode)) {
+                    if (objType !in context.visitedNodes()) {
+                        additionalObjectTypes += objType
+                    }
+                }
+                return TraversalControl.CONTINUE
+            }
+
+            val type = GraphQLTypeUtil.unwrapNonNull(node.type)
 
             val dataFetcher: DataFetcher<*> = if (mockConfig[pathForNode] != null) {
                 logger.info("Returning provided mock data for {}", pathForNode)
-                providedRoots.add(pathForNode)
+                providedRoots += pathForNode
                 getProvidedMockData(pathForNode)
             } else {
                 logger.info("Generating mock data for {}", pathForNode)
                 when (type) {
                     is GraphQLScalarType -> {
-                        DataFetcher { generateDataForScalar(type.name) }
+                        StaticDataFetcher(generateDataForScalar(type))
+                    }
+                    is GraphQLEnumType -> {
+                        StaticDataFetcher(type.values.random().value)
                     }
                     is GraphQLList -> {
-                        val wrappedType = when (type.wrappedType) {
-                            is GraphQLNamedType -> (type.wrappedType as GraphQLNamedType).name
-                            else -> return super.visitGraphQLFieldDefinition(node, context)
+                        val elementType = GraphQLTypeUtil.unwrapNonNull(type.wrappedType)
+                        if (elementType !is GraphQLNamedType) {
+                            return TraversalControl.CONTINUE
                         }
-
-                        val mockedValues = (0..faker.number().numberBetween(0, 10))
-                            .map { generateDataForScalar(wrappedType) }
-                            .toList()
-                        DataFetcher { mockedValues }
+                        val mockedValues: Collection<Any?> = when (elementType) {
+                            is GraphQLScalarType -> (0..faker.number().numberBetween(0, 10))
+                                .map { generateDataForScalar(elementType) }
+                            is GraphQLEnumType -> (0..faker.number().numberBetween(0, 3))
+                                .asSequence()
+                                .map { elementType.values.random().name }
+                                .toSet()
+                            else -> (0..faker.number().numberBetween(0, 10)).map { dummyObject(elementType) }
+                        }
+                        StaticDataFetcher(mockedValues)
                     }
-                    else -> DataFetcher { Object() }
+                    else -> StaticDataFetcher(dummyObject(type))
                 }
             }
 
-            mockFetchers[FieldCoordinates.coordinates((context?.parentNode as GraphQLObjectType).name, node?.name)] = dataFetcher
+            mockFetchers[FieldCoordinates.coordinates(parentNode, node)] = dataFetcher
         }
 
-        return super.visitGraphQLFieldDefinition(node, context)
+        return TraversalControl.CONTINUE
     }
 
-    private fun generateDataForScalar(type: String): Any {
-        return when (type) {
-            "String" -> faker.book().title()
-            "Boolean" -> faker.bool().bool()
-            "Int" -> faker.number().randomDigit()
-            "Float" -> faker.number().randomDouble(2, 0, 100000)
-            "ID" -> faker.number().digit()
-            else -> Object()
+    private fun dummyObject(type: GraphQLType): Any {
+        val displayName = GraphQLTypeUtil.simplePrint(type)
+        return object {
+            override fun toString(): String {
+                return "DummyObject{type=$displayName}"
+            }
+        }
+    }
+
+    private fun generateDataForScalar(scalarType: GraphQLScalarType): Any {
+        return when (scalarType) {
+            Scalars.GraphQLString -> faker.book().title()
+            Scalars.GraphQLBoolean -> faker.bool().bool()
+            Scalars.GraphQLInt -> faker.number().randomDigit()
+            Scalars.GraphQLFloat -> faker.number().randomDouble(2, 0, 100000)
+            Scalars.GraphQLID -> faker.number().digit()
+            else -> return dummyObject(scalarType)
         }
     }
 
     private fun getProvidedMockData(pathForNode: String?): DataFetcher<*> {
         return when (val provided = mockConfig[pathForNode]) {
             is DataFetcher<*> -> provided
-            else -> DataFetcher { provided }
+            else -> StaticDataFetcher(provided)
         }
     }
 
-    private val getPathForNode = { parents: List<GraphQLSchemaElement>?, node: GraphQLFieldDefinition? -> parents?.filterIsInstance<GraphQLFieldDefinition>()?.map { it.name }?.fold(node?.name) { n, result -> "$result.$n" } }
+    private fun getPathForNode(parents: List<GraphQLSchemaElement>, node: GraphQLFieldDefinition): String {
+        return (parents.asReversed().asSequence().filterIsInstance<GraphQLFieldDefinition>() + sequenceOf(node))
+            .map { it.name }
+            .joinToString(".")
+    }
 }

--- a/graphql-dgs-mocking/src/test/kotlin/com/netflix/graphql/mocking/MockGraphQLVisitorTest.kt
+++ b/graphql-dgs-mocking/src/test/kotlin/com/netflix/graphql/mocking/MockGraphQLVisitorTest.kt
@@ -24,17 +24,23 @@ import graphql.schema.DataFetcher
 import graphql.schema.FieldCoordinates
 import graphql.schema.GraphQLCodeRegistry
 import graphql.schema.GraphQLSchema
+import graphql.schema.StaticDataFetcher
 import graphql.schema.idl.RuntimeWiring
 import graphql.schema.idl.SchemaGenerator
 import graphql.schema.idl.SchemaParser
+import graphql.schema.idl.TypeRuntimeWiring
+import org.assertj.core.api.Assertions.LIST
+import org.assertj.core.api.Assertions.MAP
+import org.assertj.core.api.Assertions.STRING
+import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.Assertions
 import org.junit.jupiter.api.Test
 
-internal class MockGraphQLVisitorTest {
+class MockGraphQLVisitorTest {
 
     @Test
     fun generateDataForStringScalar() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -51,12 +57,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertTrue((someObject["someKey"] as String).isNotEmpty())
+        assertThat(someObject["someKey"]).asString().isNotEmpty()
     }
 
     @Test
     fun generateDataForBooleanScalar() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -73,12 +79,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(Boolean::class, someObject["someKey"]!!::class)
+        assertThat(someObject["someKey"]).isInstanceOf(Boolean::class.javaObjectType)
     }
 
     @Test
     fun generateDataForIntScalar() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -95,12 +101,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(Int::class, someObject["someKey"]!!::class)
+        assertThat(someObject["someKey"]).isInstanceOf(Int::class.javaObjectType)
     }
 
     @Test
     fun generateDataForFloatScalar() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -117,12 +123,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(Double::class, someObject["someKey"]!!::class)
+        assertThat(someObject["someKey"]).isInstanceOf(Double::class.javaObjectType)
     }
 
     @Test
     fun generateDataForIDScalar() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -139,12 +145,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(String::class, someObject["someKey"]!!::class)
+        assertThat(someObject["someKey"]).asInstanceOf(STRING).isNotBlank()
     }
 
     @Test
     fun generateDataForNonNullableString() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -161,12 +167,12 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(String::class, someObject["someKey"]!!::class)
+        assertThat(someObject["someKey"]).asInstanceOf(STRING).isNotBlank()
     }
 
     @Test
     fun generateDataForStringList() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -185,14 +191,14 @@ internal class MockGraphQLVisitorTest {
         val someObject = data["someObject"] as Map<*, *>
 
         when (val value = someObject["someKey"]!!) {
-            is List<*> -> value.forEach { Assertions.assertTrue(it is String) }
+            is List<*> -> value.forEach { assertThat(it).isInstanceOf(String::class.java) }
             else -> Assertions.fail("Returned mock is not a List")
         }
     }
 
     @Test
     fun generateDataForObject() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -214,13 +220,12 @@ internal class MockGraphQLVisitorTest {
 
         val someObject = data["someObject"] as Map<*, *>
         val someKey = someObject["someKey"] as Map<*, *>
-        Assertions.assertNotNull(someKey)
-        Assertions.assertTrue(someKey["name"] is String)
+        assertThat(someKey["name"]).asInstanceOf(STRING).isNotBlank()
     }
 
     @Test
     fun generateDataForObjectList() {
-        val mockConfig = mapOf(Pair("someObject.someKey", null))
+        val mockConfig = mapOf("someObject.someKey" to null)
 
         val schema = createSchema(
             """
@@ -242,12 +247,15 @@ internal class MockGraphQLVisitorTest {
 
         val someObject = data["someObject"] as Map<*, *>
         val myObjectArr = someObject["someKey"] as List<*>
-        Assertions.assertTrue(((myObjectArr[0] as Map<*, *>)["name"] as String).isNotBlank())
+        assertThat(myObjectArr[0]).asInstanceOf(MAP)
+            .extractingByKey("name")
+            .asInstanceOf(STRING)
+            .isNotBlank()
     }
 
     @Test
     fun providedMockData() {
-        val mockConfig = mapOf(Pair("someObject.someKey", listOf("a", "b", "c")))
+        val mockConfig = mapOf("someObject.someKey" to listOf("a", "b", "c"))
 
         val schema = createSchema(
             """
@@ -264,12 +272,13 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(listOf("a", "b", "c"), someObject["someKey"])
+        assertThat(someObject["someKey"]).asInstanceOf(LIST)
+            .containsExactly("a", "b", "c")
     }
 
     @Test
     fun providedMockDataForObject() {
-        val mockConfig = mapOf(Pair("someObject.someKey", listOf(MyObject(name = "mymockedvalue"))))
+        val mockConfig = mapOf("someObject.someKey" to listOf(MyObject(name = "mymockedvalue")))
 
         val schema = createSchema(
             """
@@ -291,15 +300,14 @@ internal class MockGraphQLVisitorTest {
 
         val someObject = data["someObject"] as Map<*, *>
         val someKeyList = someObject["someKey"] as List<*>
-        Assertions.assertNotNull(someKeyList)
-        Assertions.assertEquals(1, someKeyList.size)
-        Assertions.assertEquals("mymockedvalue", (someKeyList[0] as Map<*, *>)["name"])
+        assertThat(someKeyList).size().isEqualTo(1)
+        assertThat(someKeyList[0]).asInstanceOf(MAP).extractingByKey("name").isEqualTo("mymockedvalue")
     }
 
     @Test
     fun providedMockDataFetcherData() {
         val dataFetcher = DataFetcher { listOf("a", "b", "c") }
-        val mockConfig = mapOf(Pair("someObject.someKey", dataFetcher))
+        val mockConfig = mapOf("someObject.someKey" to dataFetcher)
 
         val schema = createSchema(
             """
@@ -316,14 +324,14 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { someObject {someKey} }", mockConfig)
 
         val someObject = data["someObject"] as Map<*, *>
-        Assertions.assertEquals(listOf("a", "b", "c"), someObject["someKey"])
+        assertThat(someObject["someKey"]).asInstanceOf(LIST).containsExactly("a", "b", "c")
     }
 
     @Test
     fun multipleMocksSimilarName() {
-        val nameFetcher = DataFetcher { "nameMock" }
-        val namesFetcher = DataFetcher { listOf("listNameMock") }
-        val mockConfig = mapOf(Pair("name", nameFetcher), Pair("names", namesFetcher))
+        val nameFetcher = StaticDataFetcher("nameMock")
+        val namesFetcher = StaticDataFetcher(listOf("listNameMock"))
+        val mockConfig = mapOf("name" to nameFetcher, "names" to namesFetcher)
 
         val schema = createSchema(
             """
@@ -337,7 +345,104 @@ internal class MockGraphQLVisitorTest {
         val data = execute(schema, "query { names }", mockConfig)
 
         val names = data["names"] as List<*>
-        Assertions.assertEquals(listOf("listNameMock"), names)
+        assertThat(names).containsExactly("listNameMock")
+    }
+
+    @Test
+    fun generateDataForEnumType() {
+        val schema = """
+            type Query {
+               animalKind: AnimalKind!
+            }
+
+            enum AnimalKind { DOG, SLOTH, BEAR, PIG }
+        """
+
+        val graphqlSchema = createSchema(schema)
+
+        val data = execute(graphqlSchema, "query { animalKind }", mapOf("animalKind" to null))
+        val animalKind = data["animalKind"] as String
+        assertThat(animalKind).isIn("DOG", "SLOTH", "BEAR", "PIG")
+    }
+
+    @Test
+    fun generateDataForListOfEnumType() {
+        val schema = """
+            type Query {
+               animalKinds: [AnimalKind!]!
+            }
+
+            enum AnimalKind { DOG, SLOTH, BEAR, PIG }
+        """
+
+        val graphqlSchema = createSchema(schema)
+
+        val data = execute(graphqlSchema, "query { animalKinds }", mapOf("animalKinds" to null))
+        val animalKind = data["animalKinds"] as List<*>
+        assertThat(animalKind).isSubsetOf("DOG", "SLOTH", "BEAR", "PIG")
+    }
+
+    @Test
+    fun generateDataForUnionType() {
+        val schema = """
+            type Query {
+               animal: Animal
+            }
+
+            union Animal = Dog | Sloth
+
+            type Dog {
+              name: String
+              barkVolume: Int
+            }
+
+            type Sloth {
+              chill: Boolean
+              speed: Int
+            }
+        """
+
+        val graphqlSchema = createSchema(schema) { builder ->
+            builder.type(TypeRuntimeWiring.newTypeWiring("Animal").typeResolver { env -> env.schema.getObjectType("Sloth") })
+        }
+
+        val data = execute(graphqlSchema, "query { animal { __typename ... on Sloth { chill speed } } }", mapOf("animal" to null, "Sloth" to null))
+        val slothData = data["animal"] as Map<*, *>
+        assertThat(slothData["chill"]).isInstanceOf(Boolean::class.javaObjectType)
+        assertThat(slothData["speed"]).isInstanceOf(Int::class.javaObjectType)
+    }
+
+    @Test
+    fun generateDataForInterfaceType() {
+        val schema = """
+            type Query {
+               animal: Animal
+            }
+
+            interface Animal {
+              speed: Int
+            }
+
+            type Dog implements Animal {
+              name: String
+              barkVolume: Int
+              speed: Int
+            }
+
+            type Sloth implements Animal {
+              chill: Boolean
+              speed: Int
+            }
+        """
+
+        val graphqlSchema = createSchema(schema) { builder ->
+            builder.type(TypeRuntimeWiring.newTypeWiring("Animal").typeResolver { env -> env.schema.getObjectType("Sloth") })
+        }
+
+        val data = execute(graphqlSchema, "query { animal { __typename ... on Sloth { chill speed } } }", mapOf("animal" to null))
+        val slothData = data["animal"] as Map<*, *>
+        Assertions.assertInstanceOf(Boolean::class.javaObjectType, slothData["chill"])
+        Assertions.assertInstanceOf(Int::class.javaObjectType, slothData["speed"])
     }
 
     private fun execute(schema: GraphQLSchema, query: String, mockConfig: Map<String, Any?>): Map<String, *> {
@@ -350,19 +455,23 @@ internal class MockGraphQLVisitorTest {
             .build()
 
         val executionResult = graphQL.execute(executionInput)
+        if (!executionResult.isDataPresent) {
+            throw AssertionError("GraphQL query failed: $executionResult")
+        }
         return executionResult.getData()
     }
 
-    private fun createSchema(schema: String): GraphQLSchema {
+    private fun createSchema(schema: String, block: (RuntimeWiring.Builder) -> Unit = {}): GraphQLSchema {
         val schemaParser = SchemaParser()
         val typeDefinitionRegistry = schemaParser.parse(schema.trimIndent())
 
-        val someObjectDf = DataFetcher { SomeObject() }
+        val someObjectDf = StaticDataFetcher(SomeObject())
 
         val codeRegistry = GraphQLCodeRegistry.newCodeRegistry()
             .dataFetcher(FieldCoordinates.coordinates("Query", "someObject"), someObjectDf)
+            .build()
 
-        val runtimeWiring = RuntimeWiring.newRuntimeWiring().codeRegistry(codeRegistry).build()
+        val runtimeWiring = RuntimeWiring.newRuntimeWiring().codeRegistry(codeRegistry).build().transform(block)
 
         val schemaGenerator = SchemaGenerator()
         return schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring)

